### PR TITLE
Add support for hierarchical list prompts

### LINKED ISSUE #412: Multiple selection with groups
**Is your feature request related to a problem? Please describe.**

It would be awesome to be able to do multiple selections with groups. Something like:

```
[ ] My group
  [ ] Item 1
  [ ] Item 2
[ ] My group 2
  [ ] Another one
  [ ] Yet another one
```

And be able to group and select in groups, etc

### PR DESCRIPTION
**Pull request by patriksvensson**
This PR introduces support for hierarchical list prompts.

There is one breaking change here. Items in multi-selection prompts can no longer be configured to be pre-selected using an index since this doesn't make sense anymore.

prompt_tree

Closes #412

### CODE REVIEW

**Review by phil-scott-78 — COMMENTED** [review]
Code looks good, but I'm going back and forth with the `Hierarchical` property name. Without actually running the code it wasn't terribly clear what would happen between setting it vs `NonHierarchical`. Hierarchical to me indicates that there is going to be a ranking, but nothing that the parents being selectedable.

I kind of like how Vuetify names these two modes "leaf" and "independent" where leaf mode you can only select the leafs vs independent which allows parents to be selected too.

**Comment by patriksvensson** [comment]
@phil-scott-78 That's some good feedback. I've added a `SelectionMode` enum that contains `Leaf` and `Independent` like Vuetify has, which I think makes sense. I've also added selection mode support for `SelectionPrompt` as well.

**Review by phil-scott-78 — APPROVED** [review]

File: examples/Console/Prompt/Program.cs [hunk 1]
```diff
@@ -11,6 +13,26 @@ public static void Main(string[] args)
                 return;
             }
 
+            var favorites = AnsiConsole.Prompt(
```
Comment (phil-scott-78) [inline examples/Console/Prompt/Program.cs:16]: I missed this when doing the first pass on the review. I suspect we didn't want this added.

Comment (patriksvensson) [reply examples/Console/Prompt/Program.cs:16]: You didn't miss it. I accidentally left it in the latest update 😅

Comment (patriksvensson) [reply examples/Console/Prompt/Program.cs]: @phil-scott-78 Removed!

**Pull request merged**